### ISSUE TITLE
Build adhoc modules into mimosa

### ISSUE THREAD

**Issue by dbashford** [dbashford/mimosa#440]
Should be able to simply add ad-hocs to the module list as normal modules.

```javascript
modules:[
  "copy",
  "jshint",
  "adhoc@module1",
  "adhoc@module2"
]
```

The assumed location would be in `scripts`. If not found in scripts, the string after `@` will be used to find the adhoc relative to the root of the project.  So...

```javascript
modules:[
  "copy",
  "jshint",
  "adhoc@adhocs/adhoc1",
  "adhoc@adhocs/adhoc2"
]
```

Thoughts anyone?

**Comment by Anachron** [comment]
Yes, I like it. On the other way, it would require mimosa to have adhoc modules implemented inside the core, else it would be kinda weird to place it inside the modules array.

**Comment by timkendrick** [comment]
I've only recently stumbled across Mimosa, but for me ad-hoc modules are the killer feature: I'd love to see it as part of the core. Just a couple of thoughts:
1. Syntax-wise, would it make more sense for ad-hoc modules to follow Node's `require()` convention of starting with a period to denote relative resources?
   
   This would allow you to write the following, which doesn't make any assumptions about a project having a `scripts` folder:
   
   ```
   modules: [
     "copy",
     "jshint",
     "./adhocs/module1",
     "./adhocs/module2"
   ]
   ```
2. Similarly to how mimosa auto-detects npm packages starting with `mimosa-`, could Mimosa also auto-detect ad-hoc modules in a certain (user-configurable) folder relative to the project root, say `./mimosa_modules` by default? I feel like this would make working with ad-hoc modules completely frictionless.

Like I say, I'm new to this project so apologies if I'm way off track here!

**Comment by dbashford** [comment]
Not off base at all, and welcome!  You picked a good time to check it out.  Lots of good stuff coming in the next ~4-6 weeks with 3.0.
1. Like the idea of the paths.  Mimosa could just detect a module is a path string string and follow it rather than the `adhoc@` syntax.  Essentially if `path.join(project root, module string)` = path to file then use it, otherwise do what it does now and try to find locally and mimosa-installed modules.  Dig it!
2. Also like the idea of auto-detection, but other things become problems if you do that.  What if you have separate run profiles and you want to use an adhoc in one profile and not the other?  We do that quite a bit.  Think we're better off having to provide the modules you want to use.  While mimosa will auto-install mimosa-modules that it cannot find, it never auto-includes them in your project (at least it hasn't for well over a year).  The flexibility/power that profiles provide are more important than skipping out on having to name `modules` I think.

**Comment by timkendrick** [comment]
Thanks for the fast response!
1. Great! Although in the example you gave, wouldn't path-string checking present problems if you happened to have a non-mimosa-related folder in your project root that happened to be named `copy`?
2. Ah OK, I misread the [install docs](http://mimosa.io/modules.html#use): I thought it was saying that adding it to the `modules` array ("easiest way") was an _alternative_ to the "npm install" method (it says that the npm-installed modules would "take precedence over any modules installed into other scopes", which I assumed meant the `modules` array).

Sorry for the confusion, goes to show I should probably get some experience using this before throwing in my two cents. Can't wait for 3.0, looking forward to using Mimosa on my next project...

**Comment by dbashford** [comment]
For the path checking, yeah, could be a problem, but not for folders.  It would assume `copy.js` and possibly even check if the file was a file before requiring it in.

If you did have a `copy.js` sitting in the root of your project, that could present problems.  But if Mimosa doesn't find any of the functions it expects (`register`, `registerCommand`) it would just carry on after possibly messaging to the user that an adhoc-module doesn't expose any of the expected methods.

For the install thing, if you have it installed via `npm install`, and then add it to the `modules` array, then Mimosa will use it.  However, because Mimosa will auto-install modules if it cannot find them, by skipping the `npm install` part, you essentially accomplish the same thing.  That is why the just-add-it-to-modules is easier.
